matrix/common/request_cache.py

```diff
@@ -0,0 +1,88 @@
+import hashlib
+import typing
+
+from matrix.common.dynamo_handler import DynamoHandler
+from matrix.common.logging import Logging
+
+logger = Logging.get_logger(__name__)
+
+# Special value representing no hash. Dynamo doesn't like empty fields.
+NULL_REQUEST_HASH = "null"
+
+
+class RequestNotFound(Exception):
+    """Exception for request ids missing the the request cache table."""
+    pass
+
+
+class RequestCache(object):
+    """Interface to request caching and hashing.
+
+    Parameters
+    ----------
+    request_id : uuid
+        id of the request to calculate and look up hashed for
+
+    Methods
+    -------
+    initialize()
+        Create an empty entry for the request id in the cache table
+
+    retrieve_hash()
+        Look up the hash associated with the request id
+
+    set_hash(bundle_fqids, format)
+        Calculate, set, and return the hash value for the request id based on
+        the request's bundles and format.
+    """
+
+    def __init__(self, request_id: str) -> None:
+        self._request_id = request_id
+        self._dynamo_handler = DynamoHandler()
+
+    @staticmethod
+    def _hash_request(bundle_fqids: typing.List[str], format_: str) -> str:
+        """Calculate the hash of a request defined by fqids and a requested format."""
+        hash_obj = hashlib.sha256()
+
+        for fqid in sorted(bundle_fqids):
+            hash_obj.update(fqid.encode())
+        if format_:
+            hash_obj.update(format_.encode())
+
+        return hash_obj.hexdigest()
+
+    def initialize(self) -> None:
+        """Initialize the request id in the request cache table.
+
+        Sets a null hash value.
+        """
+        self._dynamo_handler.write_request_hash(self._request_id, NULL_REQUEST_HASH)
+
+    def retrieve_hash(self) -> typing.Union[str, None]:
+        """Look up the hash value for the request.
+
+        Returns None if the request has been initialized but the hash has not yet
+        been set.
+
+        Raises RequestNotFound if the request id isn't in the cache table.
+        """
+        request_hash = self._dynamo_handler.get_request_hash(self._request_id)
```

Comment:
Use `get_table_item` instead?

Reply:
I think this only works if I parametrize `get_table_item` a lot more. The key name is different for the cache table, and I kind of want different things to happen when ids aren't found.

Reply:
Ohh the key names are different. Gotcha